pysingfel/detector/base.py

```diff
@@ -246,9 +270,59 @@ def get_photons(self, particle, device=None):
                 "Device option is deprecated. "
                 "Everything now runs on the GPU.")
 
-        raw_data = self.get_pattern_without_corrections(particle=particle)
+        raw_data = self.get_pattern_without_corrections(particle=particle,return_type="intensity")
         return self.add_correction_and_quantization(raw_data)
 
+
+    def maxRadius(self, particles):
+        radius_current = 0
+        for particle in particles:
+            print (particle, '->', particles[particle])
+            radius_arr = particle.atom_pos - np.mean(particle.atom_pos, axis=0)
+            for row in radius_arr:
+                radius = np.sqrt(row[0]**2+row[1]**2+row[2]**2)
+                if radius > radius_current:
+                    radius_current = radius
+        radius_max = radius_current
+        return radius_max
+
+
+    def distribute(self, particles, beam_focus_radius, jet_radius): #beam_focus_radius = 10e-6 #jet_radius = 1e-4
+        state = []
+        for particle in particles:
+            for count in range(particles[particle]):
+                state.append(particle)
+        radius_max = self.maxRadius(particles)
+        N = sum(particles.values()) # total number of particles
+        coords = np.zeros((N,3)) # initialize N*3 array
+        # generate N*3 random positions
+        for i in range(N):
+            coords[i,0] = beam_focus_radius*np.random.uniform(-1, 1)
+            coords[i,1] = beam_focus_radius*np.random.uniform(-1, 1)
+            coords[i,2] = jet_radius*np.random.uniform(-1, 1)
+        # calculate N*N distance matrix
+        dist_matrix = distance.cdist(coords, coords, 'euclidean')
+        # collision detection check (<2 maxRadius)
+        collision = dist_matrix < 2*radius_max
+        checkList = [collision[i][j] for i in range(N) for j in range(N) if j > i]
+        if any(item == True for item in checkList):
+            self.distribute(particles, beam_focus_radius, jet_radius)
+        return state, coords
+
+
+    def get_fxs_photons(self, particles, beam_focus_radius, jet_radius, device=None):
+        raw_data = None
+        state, coords = self.distribute(particles, beam_focus_radius, jet_radius)
+        for i in range(len(state)):
+            this_data = self.get_pattern_without_corrections(particle=state[i], return_type="complex_field")
+            this_data *= np.exp(1j * 2 * np.pi * 1e-10 * np.dot(self.pixel_position_reciprocal, coords[i]))
```

Comment:
While preparing the pull request, I changed the particule transposition into the addition of a phase-shift.
I pretty sure the `1e-10` is needed but I am not 100% confident about the `2 * np.pi`.
I think this is fine for now, but we'll need to add some tests.

Reply:
@irischang020 could you write unit tests that checks the correctness of this code

Reply:
@irischang020, my tests on small particles seem to show that the exponential above is the correct formula, only the coefficient is `1j * 1e-10` without the `2 * np.pi`. Can you check again?

Reply:
I added the add_phase_shift function in PR #32. It seems that the 1e-10 shouldn't be there if the coords are in SI units.